File: SearchA2DMatrixII/Solution.js
var searchRow = function(row, target) {
    var l = 0;
    var r = row.length;
    while (l < r) {
        m = l + ((r - l) >> 1);
        if (row[m] < target) l = m + 1;
        else if (row[m] == target) return true;
        else r = m;
    }
    return false;
};

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    var l = 0;
    var r = matrix.length;
    while (l < r) {
        m = l + ((r - l) >> 1);
        if (matrix[m][0] <= target) l = m + 1;
        else r = m;
    }
    for (var i = l - 1; i >= 0; i--) {
        if (searchRow(matrix[i], target)) return true;
    }
    return false;
};

File: SearchA2DMatrixII/Solution.rb
def search_row(row, target)
    l = 0
    r = row.length
    while l < r do
        m = l + (r - l) / 2
        if row[m] < target then
            l = m + 1
        elsif row[m] == target then
            return true
        else
            r = m
        end
    end
    return false
end

# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
def search_matrix(matrix, target)
    l = 0
    r = matrix.length
    while l < r do
        m = l + (r - l) / 2
        if matrix[m][0] <= target then
            l = m + 1
        else
            r = m
        end
    end
    (l - 1).downto(0) do | i |
        return true if search_row(matrix[i], target)
    end
    return false
end
